### PR TITLE
DAOS-5288 bio: Implement bio_replace_dev()

### PR DESCRIPTION
**Pull request by NiuYawei**
Implemented bio_replace_dev() to replace a faulty device with a
new hot plugged device, or revive a faulty device (usually used
during testing). The function will be called by dmg device replace
command handler, and being scheduled on init xstream.

Changes to SMD for implementing bio_replace_dev():
- Added blob size in SMD pool table (durable format change);
- Added smd_dev_replace() API;

Signed-off-by: Niu Yawei <yawei.niu@intel.com>

### CODE REVIEW

File: src/bio/smd/tests/smd_ut.c [hunk 1]
```diff
@@ -244,9 +251,95 @@ ut_pool(void **state)
...
+	/* Replace dev1 with dev3 without marking dev1 as faulty */
+	rc = smd_dev_replace(dev_id1, dev_id3, &pool_list);
+	assert_int_equal(rc, -DER_INVAL);
+
+	rc = smd_dev_set_state(dev_id1, SMD_DEV_FAULTY);
+	assert_int_equal(rc, 0);
+
+	/* Replace dev1 with dev2 */
+	rc = smd_dev_replace(dev_id1, dev_id2, &pool_list);
+	assert_int_equal(rc, -DER_INVAL);
```
Comment (tanabarr) [inline src/bio/smd/tests/smd_ut.c:298]: maybe a comment to describe why this fails, it looks like it's because the target id list doesn't match between `dev_id1` & `dev_id2`. rc `-DER_MISMATCH` would also be more intuitive in this case.

**Review by tanabarr — APPROVED** [review]
I can't attest to the correctness of the iov related code as I'm not familiar enough with it , but the rest of the logic in the PR looks sound. Request for a comment expansion in the unit test but that's not worth a re-push.  Observation that the unit tests are a little hard to follow due to the amount of shared global state and might be improved by using setup routines to set the initial state at the beginning of each unit test case.

File: src/bio/smd/tests/smd_ut.c (hunk 1 above)
Comment (NiuYawei) [reply src/bio/smd/tests/smd_ut.c:298]: Right, I'd have left a comment here. The 3 tests in smd_ut.c are actually dependent, device test -> pool test -> replace test. There are two devices (dev1 and dev2) are inserted in SMD in device test, and replacing a device with another device in SMD (which means the device is used by DAOS) isn't allowed, so that's why it should return -DER_INVAL.

An faulty device can only be replaced by a new device not being used by DAOS. (not present in SMD)

**Comment by NiuYawei** [comment]
Leap 15 build failure and 5 known CI failures: pool evict, core dfs, 2 aggregation basic are due to DAOS-5610. nvme object is due to DAOS-4943.

**Review by tanabarr — APPROVED** [review]

File: src/bio/smd/tests/smd_ut.c (hunk 1 above)
Comment (jolivier23) [reply src/bio/smd/tests/smd_ut.c:298]: DER_MISMATCH is documented to mean a mismatch in pool group version.   We should either change that or create new errors for new issues.

File: src/bio/smd/tests/smd_ut.c [hunk 2]
```diff
@@ -244,9 +251,95 @@ ut_pool(void **state)
...
+	rc = smd_dev_replace(dev_id1, dev_id3, &pool_list);
+	assert_int_equal(rc, 0);
+
+	/* Verify device after replace */
+	D_INIT_LIST_HEAD(&dev_list);
+	rc = smd_dev_list(&dev_list, &dev_cnt);
+	assert_int_equal(rc, 0);
+	assert_int_equal(dev_cnt, 2);
+
+	d_list_for_each_entry_safe(dev_info, tmp_dev, &dev_list, sdi_link) {
```
Comment (jolivier23) [inline src/bio/smd/tests/smd_ut.c:310]: For lists where you are deleting the list, it's a bit more natural to use d_list_pop_entry in a while loop (e.g. while ((dev_info = d_list_entry_pop(...)) != NULL) { )

**Review by jolivier23 — APPROVED** [review]
Looks fine to me

**Review by sydneyvanda — APPROVED** [review]
overall looks good, once the work to replace device with dmg is ready we can test more thoroughly

File: src/bio/bio_recovery.c [hunk 3]
```diff
@@ -477,6 +466,43 @@ bio_bs_state_set(struct bio_blobstore *bbs, enum bio_bs_state new_state)
...
+on_normal(struct bio_blobstore *bbs)
+{
+	struct bio_bdev	*bdev = bbs->bb_dev;
+	int		 tgt_ids[BIO_XS_CNT_MAX];
+	int		 tgt_cnt, i, rc;
+
+	/*
+	 * Trigger auto reint only when faulty is replaced by new hot
+	 * plugged device. See comments in bio_replace_dev().
+	 */
```
[the remark this answers is not in the document]
Comment (NiuYawei) [reply src/bio/bio_recovery.c:479]: If it's a normal device replugged, it's targets are never being marked as DOWN, we don't support that kind of "reint" (reint a target not marked as DOWN before).

File: src/bio/smd/smd_internal.h [hunk 4]
```diff
@@ -45,7 +45,8 @@ POBJ_LAYOUT_END(smd_md_layout);
 
 #define SMD_DF_MAGIC		0x5eaf00d
 
-#define SMD_DF_VER_1		1
+/* The lowest supported version */
+#define SMD_DF_VER_1		2
```
[the remark this answers is not in the document]
Comment (NiuYawei) [reply src/bio/smd/smd_internal.h:49]: Yes, a new version.

File: src/bio/smd/tests/smd_ut.c (hunk 1 above)
Comment (NiuYawei) [reply src/bio/smd/tests/smd_ut.c:298]: Ok, I'll change it to some other error code.

Comment (NiuYawei) [reply src/bio/smd/tests/smd_ut.c:298]: I mean in follow-on patch.

**Comment by NiuYawei** [comment]
> Leap 15 build failure and 5 known CI failures: pool evict, core dfs, 2 aggregation basic are due to DAOS-5610. nvme object is due to DAOS-4943.

Can we land it? I'll change the DER_MISMATCH error code in follow-on patch.
